torch/csrc/jit/codegen/cuda/ir_cloner.cpp

```diff
@@ -0,0 +1,132 @@
+
+#include <torch/csrc/jit/codegen/cuda/ir_cloner.h>
+#include <torch/csrc/jit/codegen/cuda/fusion.h>
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+Statement* IrCloner::clone(const Statement* statement) {
+  if (statement == nullptr) {
+    return nullptr;
+  }
+
+  // Have we already cloned this node?
+  const auto it = clones_map_.find(statement);
+  if (it != clones_map_.end()) {
+    return it->second;
+  } else {
+    // Clone the new node, saving/restoring this->clone_
+    // since the cloning can be reentrant
+    auto saved_clone = clone_;
+    handle(statement);
+    auto new_node = clone_;
+    clone_ = saved_clone;
+
+    // The base cloning constructor (Statement) should have
+    // registered the new node
+    TORCH_CHECK(new_node != nullptr);
+    TORCH_CHECK(clones_map_[statement] == new_node);
```

Comment:
Should these be TORCH_INTERNAL_ASSERT? Can you add a comment as to why it might have failed?

Reply:
I'm not sold on the TORCH_INTERNAL_ASSERT vs TORCH_CHECK distinction, so I prefer to use TORCH_CHECK by default, and change them when explicitly requested. Such in this case - fixed.

These are paranoid checks and I contemplated getting rid of the IrCloner::clone_ completely.

Reply:
I would rather err on the side of paranoid checks.